src/messages/MMDSQuiesceDbListing.h

```diff
@@ -0,0 +1,63 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software 
+ * Foundation.  See file COPYING.
+ * 
+ */
+
+
+#pragma once
+
+#include "messages/MMDSOp.h"
+#include "mds/QuiesceDbEncoding.h"
+
+class MMDSQuiesceDbListing final : public MMDSOp {
+public:
+  mds_gid_t gid;
+  mutable QuiesceDbListing db_listing;
```

Comment:
I see your concern, let's figure it out.

My problem is that messages tend to keep the data twice: once in the serialized form in the payload, and once in the unpacked form in a member like this one. I want to consume that unpacked form.

Please point me to the flow where a message can be retried during the receipt and where my current approach will result in empty message contents.

Reply:
There may not be an **existing** case with your current code where the message is retried. That's not relevant however. The point is that we require in new code that messages can be retried without a past try invalidating its state.

Reply:
Please clarify "retried".

Reply:
```
batrick@menzoberranzan ~/ceph/src/mds$ git grep RetryMessage
Locker.cc:      mds->wait_for_replay(new C_MDS_RetryMessage(mds, m));
Locker.cc:    mds->wait_for_replay(new C_MDS_RetryMessage(mds, m));
Locker.cc:      mdcache->wait_replay_cap_reconnect(m->get_ino(), new C_MDS_RetryMessage(mds, m));
Locker.cc:    head_in->add_waiter(CInode::WAIT_UNFREEZE, new C_MDS_RetryMessage(mds, m));
Locker.cc:    mds->wait_for_replay(new C_MDS_RetryMessage(mds, m));
Locker.cc:                       new C_MDS_RetryMessage(mds, m));
MDBalancer.cc:    mds->mdcache->wait_for_open(new C_MDS_RetryMessage(mds, m));
MDCache.cc:      mds->wait_for_resolve(new C_MDS_RetryMessage(mds, m));
MDCache.cc:      mds->wait_for_rejoin(new C_MDS_RetryMessage(mds, strong));
MDCache.cc:      CF_MDS_RetryMessageFactory cf(mds, m);
MDCache.cc:      mdr->more()->waiting_for_finish.push_back(new C_MDS_RetryMessage(mds, req));
MDCache.cc:      mds->wait_for_replay(new C_MDS_RetryMessage(mds, dis));
MDCache.cc:     cur->add_waiter(CInode::WAIT_UNFREEZE, new C_MDS_RetryMessage(mds, dis));
MDCache.cc:      curdir->add_waiter(CDir::WAIT_UNFREEZE, new C_MDS_RetryMessage(mds, dis));
MDCache.cc:       curdir->fetch(dname, snapid, new C_MDS_RetryMessage(mds, dis),
MDCache.cc:     dn->lock.add_waiter(SimpleLock::WAIT_RD, new C_MDS_RetryMessage(mds, dis));
MDCache.cc:     dnl->get_inode()->add_waiter(CDir::WAIT_UNFREEZE, new C_MDS_RetryMessage(mds, dis));
MDCache.cc:      CF_MDS_RetryMessageFactory cf(mds, m);
MDCache.cc:     open_remote_dirfrag(in, df.frag, new C_MDS_RetryMessage(mds, m));
MDSRank.h:    friend class C_MDS_RetryMessage;
MDSRank.h:class C_MDS_RetryMessage : public MDSInternalContext {
MDSRank.h:  C_MDS_RetryMessage(MDSRank *mds, const cref_t<Message> &m)
MDSRank.h:class CF_MDS_RetryMessageFactory : public MDSContextFactory {
MDSRank.h:  CF_MDS_RetryMessageFactory(MDSRank *mds, const cref_t<Message> &m)
MDSRank.h:    return new C_MDS_RetryMessage(mds, m);
MDSTableClient.cc:      mds->wait_for_resolve(new C_MDS_RetryMessage(mds, m));
Migrator.cc:      mds->wait_for_any_client_connection(new C_MDS_RetryMessage(mds, m));
ScrubStack.cc:  gather.set_finisher(new C_MDS_RetryMessage(mdcache->mds, m));
Server.cc:      mds->enqueue_replay(new C_MDS_RetryMessage(mds, m));
Server.cc:      mds->wait_for_active(new C_MDS_RetryMessage(mds, m));
Server.cc:    mds->wait_for_replay(new C_MDS_RetryMessage(mds, m));
Server.cc:      mds->wait_for_active(new C_MDS_RetryMessage(mds, m));
Server.cc:      mds->wait_for_replay(new C_MDS_RetryMessage(mds, m));
Server.cc:    mds->wait_for_reconnect(new C_MDS_RetryMessage(mds, m));
Server.cc:    mdcache->wait_for_open(new C_MDS_RetryMessage(mds, req));
Server.cc:    mds->wait_for_replay(new C_MDS_RetryMessage(mds, m));
Server.cc:    mds->wait_for_active(new C_MDS_RetryMessage(mds, m));
Server.cc:    mds->wait_for_replay(new C_MDS_RetryMessage(mds, m));
```

Reply:
I removed the members to avoid the duplication of the structures in memory; Instead, I'm encoding and decoding on demand. Hence I consider this thread resolved, please see the latest code

Reply:
You're losing the protections built into `decode_payload` which catch decoding errors. I'd suggest you find a way to preserve that. Probably you can build a generic `Message::decode_member` method that takes a lambda (so you can decide which method in your message to call, in this case `encode_payload_from`.

Reply:
The protection is now in `MDSRank::quiesce_dispatch`, see https://github.com/ceph/ceph/pull/54485/files#diff-d4d03b797940276595f7cad04a78b63c56440ee19f3c5c5336a5994a5a2217fcR399